my-megaparsec-project/src/Optimizer.hs

```diff
@@ -0,0 +1,192 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Optimizer
+  ( checkConstExpr
+  , optConstExpr
+  , algebraicOpt
+  , canEvalStmt
+  , optStmt
+  , modifiedVars
+  , optimizeWithEmptySubst
+  , getVarsInExpr
+  , canEvalExpr
+  ) where
+
+import qualified Data.Map as M
+import Data.Text (Text)
+import qualified Data.Set as S
+import Control.Monad.State
+
+import Expression
+import Statement
+
+-- | Returns True if the expression can be evaluated at compile time
+checkConstExpr :: Expr -> Bool
+checkConstExpr (IntLit _) = True
+checkConstExpr (VarLit _) = False
+checkConstExpr (Sum a b) = checkConstExpr a && checkConstExpr b
+checkConstExpr (Sub a b) = checkConstExpr a && checkConstExpr b
+checkConstExpr (Prod a b) = checkConstExpr a && checkConstExpr b
+checkConstExpr (Div a b) = checkConstExpr a && checkConstExpr b && (b /= IntLit 0)
+checkConstExpr (Neg a) = checkConstExpr a
+
+-- | Optimizes a constant expression by evaluating it at compile time
+optConstExpr :: Expr -> Expr
+optConstExpr expr
+  | checkConstExpr expr = IntLit (evalExpr expr M.empty)
+  | otherwise = case expr of
+      IntLit n -> IntLit n
+      VarLit v -> VarLit v
+      Sum a b -> Sum (optConstExpr a) (optConstExpr b)
+      Sub a b -> Sub (optConstExpr a) (optConstExpr b)
+      Prod a b -> Prod (optConstExpr a) (optConstExpr b)
+      Div a b -> Div (optConstExpr a) (optConstExpr b)
+      Neg a -> Neg (optConstExpr a)
+
```

Comment:
I think you can avoid checking that a subexpression is constant and do this in one traversal